src/Controller/LazyControllerAbstractFactory.php

```diff
@@ -0,0 +1,177 @@
+ *     ],
+ * ],
+ * </code>
+ *
+ * Or as a factory, mapping a controller class name to it:
+ *
+ * <code>
+ * 'controllers' => [
+ *     'factories' => [
+ *         MyControllerWithDependencies::class => LazyControllerAbstractFactory::class,
+ *     ],
+ * ],
+ * </code>
+ *
+ * The latter approach is more explicit, and also more performant.
+ *
+ * The factory has the following constraints/features:
+ *
+ * - A parameter named `$config` typehinted as an array will receive the
+ *   application "config" service (i.e., the merged configuration).
+ * - Parameters type-hinted against array, but not named `$config` will
+ *   be injected with an empty array.
+ * - Scalar parameters will be resolved as null values.
+ * - If a service cannot be found for a given typehint, the factory will
+ *   raise an exception detailing this.
+ * - Some services provided by Zend Framework components do not have
+ *   entries based on their class name (for historical reasons); the
+ *   factory contains a map of these class/interface names to the
+ *   corresponding service name to allow them to resolve.
+ *
+ * `$options` passed to the factory are ignored in all cases, as we cannot
+ * make assumptions about which argument(s) they might replace.
+ */
+class LazyControllerAbstractFactory implements AbstractFactoryInterface
+{
+    /**
+     * Maps known classes/interfaces to the service that provides them; only
+     * required for those services with no entry based on the class/interface
+     * name.
+     *
+     * Extend the class if you wish to add to the list.
+     *
+     * @var string[]
+     */
+    protected $aliases = [
+        ConsoleAdapterInterface::class  => 'ConsoleAdapter',
+        FilterPluginManager::class      => 'FilterManager',
+        HydratorPluginManager::class    => 'HydratorManager',
+        InputFilterPluginManager::class => 'InputFilterManager',
+        LogFilterManager::class         => 'LogFilterManager',
+        LogFormatterManager::class      => 'LogFormatterManager',
+        LogProcessorManager::class      => 'LogProcessorManager',
+        LogWriterManager::class         => 'LogWriterManager',
+        SerializerAdapterManager::class => 'SerializerAdapterManager',
+        ValidatorPluginManager::class   => 'ValidatorManager',
+    ];
+
+    /**
+     * {@inheritDoc}
+     *
+     * @return DispatchableInterface
+     */
+    public function __invoke(ContainerInterface $container, $requestedName, array $options = null)
+    {
+        $reflectionClass = new ReflectionClass($requestedName);
+
+        if (null === ($constructor = $reflectionClass->getConstructor())) {
+            return new $requestedName();
+        }
+
+        $reflectionParameters = $constructor->getParameters();
+
+        if (empty($reflectionParameters)) {
+            return new $requestedName();
+        }
+
+        $parameters = array_map(function (ReflectionParameter $parameter) use ($container, $requestedName) {
```

Comment:
I read about performance `array_map` and `foreach` and looks a little bit slower. `use` also decrease readability for me. Maybe `foreach` will be better?